src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,669 @@
+
+    let chunks = entry.data.chunks(max_record_size);
+    let num_chunks = chunks.len();
+    chunks
+        .enumerate()
+        .map(|(i, chunk)| Record {
+            version: VERSION,
+            tp: record_type(i, num_chunks),
+            data: chunk.to_vec(),
+            entry_id: entry.id,
+            ns: entry.ns.clone(),
+            checksum: crc32fast::hash(chunk),
+        })
+        .collect()
+}
+
+pub fn maybe_emit_entry(
+    record: Record,
+    ns: &NamespaceImpl,
+    entry_records: &mut HashMap<EntryId, Vec<Record>>,
+) -> Result<Option<EntryImpl>> {
+    let mut entry = None;
+    match record.tp {
+        RecordType::Full => {
+            entry = Some(EntryImpl::try_from(vec![record])?);
+        }
+        RecordType::Last => {
+            // There must have a sequence prefix before a Last record is read.
+            let mut records =
+                entry_records
+                    .remove(&record.entry_id)
+                    .context(IllegalSequenceSnafu {
+                        error: "Missing sequence prefix",
+                    })?;
+            records.push(record);
+            entry = Some(EntryImpl::try_from(records)?);
+        }
+        _ => {
+            entry_records
+                .entry(record.entry_id)
+                .or_default()
+                .push(record);
+        }
+    }
+
+    // Filters entries by namespace.
+    if let Some(entry) = entry
+        && &entry.ns == ns
+    {
+        return Ok(Some(entry));
+    }
+    Ok(None)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Implements some utility methods for testing.
+    impl Default for Record {
+        fn default() -> Self {
+            Self {
+                version: VERSION,
+                tp: RecordType::Full,
+                ns: NamespaceImpl {
+                    region_id: 0,
+                    topic: "greptimedb_wal_topic".to_string(),
+                },
+                data: Vec::new(),
+                entry_id: 0,
+                checksum: 0,
+            }
+        }
+    }
+
+    impl Record {
+        /// Overrides tp.
+        fn with_tp(&self, tp: RecordType) -> Self {
+            Self { tp, ..self.clone() }
+        }
+
+        /// Overrides data with the given data.
+        fn with_data(&self, data: &[u8]) -> Self {
+            Self {
+                data: data.to_vec(),
+                ..self.clone()
+            }
+        }
+
+        /// Overrides entry id.
+        fn with_entry_id(&self, entry_id: EntryId) -> Self {
+            Self {
+                entry_id,
+                ..self.clone()
+            }
+        }
+
+        /// Overrides namespace.
+        fn with_ns(&self, ns: NamespaceImpl) -> Self {
+            Self { ns, ..self.clone() }
+        }
+
+        /// Overrides checksum.
+        fn with_checksum(&self, checksum: u32) -> Self {
+            Self {
+                checksum,
+                ..self.clone()
+            }
+        }
+    }
```

Comment:
Since the cloning only performs during testing, I think it's acceptable. I've tried derive_builder and I found it will make codes more complicated compared to the existing implementation.